README.md
# QuTeDB — The QUBIC Test Database

[![Build Status](https://travis-ci.org/ziotom78/qutedb.svg?branch=master)](https://travis-ci.org/ziotom78/qutedb)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![codecov](https://codecov.io/gh/ziotom78/qutedb/branch/master/graph/badge.svg)](https://codecov.io/gh/ziotom78/qutedb)
[![Go Report Card](https://goreportcard.com/badge/github.com/ziotom78/qutedb)](https://goreportcard.com/report/github.com/ziotom78/qutedb)

## Requirements

- [Go](https://golang.org) version 1.16 or newer

The program depends on the following libraries:

- [Fitsio](https://github.com/astrogo/fitsio), to read and write FITS files
- [go-chart](https://github.com/wcharczuk/go-chart), to produce charts
- [GORM](https://github.com/jinzhu/gorm), an ORM
- [go-sqlite3](http://mattn.github.io/go-sqlite3), a database driver for [SQLite](https://www.sqlite.org/index.html)
- [go.uuid](https://github.com/satori/go.uuid), for generating random UUIDs
- [Logrus](https://github.com/sirupsen/logrus), for structured logging
- [simple-scrypt](https://github.com/elithrar/simple-scrypt), for encrypting passwords
- [Viper](https://github.com/spf13/viper), for reading configuration files

Note that `go-sqlite3` is a *cgo* package, which means that you must have *gcc*
in your path before downloading and compiling this library. This is usually the
case on UNIX systems; if you are running Windows, you have to install *gcc*
first (e.g., using the [scoop](https://scoop.sh/) package manager).


## Installation

To download all the dependencies, build, and install the package in your
`$GOPATH/bin` directory, run this command:

    go build ./...
    go install ./...
    
This will add the two executables `qutedb` and `createqdbcfg` in the
`bin` directory of the `$GOPATH` folder.

You must now create a configuration file. The `go install` command you
just issued installed a small script called `createqdbcfg`. Run it,
and it will print a default configuration to `stdout`. Save it into a
file named `config.json`; this file can be kept either in the current
directory or in `$HOME/.qutedb/`. Run `createqdbcfg --help` to get an
help of the commands.


## Configuration

All the configuration is specified through a file, `config.json`, which is read
from any of the following locations (in this order):

1. The directory from where `qutedb` was run;
2. The directory `.qutedb` within the home folder (i.e., `$HOME` on UNIX
   systems, `%USERPROFILE%` on Windows);
3. The directory `/etc/qutedb` (only on UNIX systems)

Use the program `createqdbcfg` to create a skeleton for this file; use
`createqdbcfg --help` to get an help of a few parameters you can set
from the command line.

Here is an example of configuration file:

`````json
{
    "log_format": "json",
    "port_number": 80
}
`````

The following table contains all the configuration parameters that can be used
in `config.json`. A few of them have sensible defaults, if no value is provided.

| *Parameter*  | *Default* | *Meaning* |
|--------------|-----------|-----------|
| `cookie_hash_key` | None | Hash key used to encode session cookies. It must be encoded using base64 encoding, and the unencoded string should be 32 or 64 characters long |
| `cookie_block_key` | None | Block key used to encode session cookies. It must be encoded using base64 encoding, and the unencoded string should be 32 or 64 characters long |
| `log_format` | `"text"`    | Format of log messages. Possible values are `"text"` and `"json"` |
| `log_output` | `"-"` | File where to write log messages. If equal to `"-"`, write to stderr; if `"--"`, write to stdout |
| `log_level` | It depends    | Logging level. Possible values are `"error"`, `"warning"`, `"info"`, and `"debug"`, in increasing order of verbosity. The default is `"info"`, unless development mode is turned on |
| `port_number` | `8080`    | Socket port number used for publishing the API and the site |
| `read_timeout` | 15 | Timeout for HTTP read operations, in seconds |
| `static_path` | `static` | Path to the directory containing static files (e.g., images) to serve |
| `server_name` | `127.0.0.1` | Name of the server (e.g., `www.example.com`) |
| `repository_path` | `.` | Path to the folder that contains the QUBIC test data |
| `write_timeout` | 60 | Timeout for HTTP write operations, in seconds |

The following environment variables are recognized and take precedence over the
corresponding keys in `config.json`:

| *Environment variable* | *Key in `config.json`* |
|------------------------|------------------------|
| `QUTEDB_PORT_NUMBER`   | `port_number`          |
| `QUTEDB_SERVER_NAME`   | `server_name`          |
| `QUTEDB_READ_TIMEOUT`  | `read_timeout`         |
| `QUTEDB_WRITE_TIMEOUT` | `write_timeout`        |

## Authentication

The code uses the "scrypt" encryption algorithm to hash users'
passwords. See the article [Do not use
sha256crypt/sha512crypt](https://pthree.org/2018/05/23/do-not-use-sha256crypt-sha512crypt-theyre-dangerous/)
for the reasons behind this choice.

The first time the program is started, it will create a new superuser
with name `admin@localhost` and password `changeme`. Please **change
the password immediately**!

## License

This code is released under the MIT license. See the file LICENSE for more details.


CHANGELOG.md
# HEAD

# 0.5.3

- Fix a typo in the "Acqusition" page [#23](https//github.com/ziotom78/qutedb/pull/23)

# 0.5.2

- Use a sounder way to build and send ZIP files through HTTP [#22](https//github.com/ziotom78/qutedb/pull/22)

# 0.5.1

- Fix issue [#20](https://github.com/ziotom78/qutedb/pull/20) ([#21](https://github.com/ziotom78/qutedb/pull/21))

- Add two columns for calibrator files in the index and acquisition pages ([#19](https://github.com/ziotom78/qutedb/pull/19))

# 0.5.0

- Include calibration configuration and data files in the database ([#18](https://github.com/ziotom78/qutedb/pull/18))

# 0.4.1

- Be more tolerant when multiple files match a mask ([#17](https://github.com/ziotom78/qutedb/pull/17))

# 0.4.0

- Upgrade broken dependencies and support Go modules ([#15](https://github.com/ziotom78/qutedb/pull/15))

# 0.3.0

- Allow the user to download ZIP files ([#9](https://github.com/ziotom78/qutedb/pull/9))
- Generate Python code for [qutepy](https://github.com/ziotom78/qutepy) ([#10](https://github.com/ziotom78/qutedb/pull/10))

# 0.2.1

- Properly handle housekeeping files ([#8](https://github.com/ziotom78/qutedb/pull/8))

# 0.2.0

- Force download of FITS files, instead of loading them manually
- Implement dedicated page for tests

# 0.1.1

- Fix bug #2
- Fix bug #1
- Use [bootstrap-table](https://bootstrap-table.com/) for the list of tests
- The default block size for cryptographic hashes has been changed from 64 to 32, in
  order to make it compatible with AES encryption (which does not support 64 bytes)

# 0.1.0

- First release


config.go
/*
The MIT License

Copyright (c) 2018 Maurizio Tomasi

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

package qutedb

import (
	"encoding/base64"
	"fmt"
	"runtime"

	"github.com/spf13/viper"
)

// A Configuration type contains all the settings loaded from environment
// variables and configuration files. It is initialized using the Viper library.
type Configuration struct {
	// Full path of the file containing the configuration
	ConfigurationFileName string `json:"-"`

	DatabaseFile string `json:"database_file"`

	LogFormat string `json:"log_format"`
	LogLevel  string `json:"log_level"`
	LogOutput string `json:"log_output"`

	PortNumber int    `json:"port_number"`
	ServerName string `json:"server_name"`

	ReadTimeout  int64 `json:"read_timeout"`
	WriteTimeout int64 `json:"write_timeout"`

	StaticPath string `json:"static_path"`

	RepositoryPath string `json:"repository_path"`

	CookieHashKey  []byte `json:"cookie_hash_key"`
	CookieBlockKey []byte `json:"cookie_block_key"`
}

// configureViper sets up the Viper library so that it can read the
// configuration file from a variety of locations.
func configureViper() {
	// Set a number of default values for configuration parameters
	viper.SetDefault("database_file", "db.sqlite3")
	viper.SetDefault("log_format", "text")
	viper.SetDefault("log_output", "-")
	viper.SetDefault("log_level", "info")
	viper.SetDefault("port_number", 8080)
	viper.SetDefault("server_name", "127.0.0.1")
	viper.SetDefault("static_path", "static")
	viper.SetDefault("repository_path", ".")
	viper.SetDefault("read_timeout", 15)
	viper.SetDefault("write_timeout", 60)

	// Bind environment variables to configuration parameters
	viper.SetEnvPrefix("qubicdb")
	viper.BindEnv("port_number")
	viper.BindEnv("server_name")
	viper.BindEnv("read_timeout")
	viper.BindEnv("write_timeout")

	// Set where to look for the configuration file
	viper.SetConfigName("config")
	viper.SetConfigType("json")

	viper.AddConfigPath(".")
	viper.AddConfigPath("$HOME/.qutedb/")
	if runtime.GOOS != "windows" {
		viper.AddConfigPath("/etc/qutedb")
	}

	// Read the configuration
	err := viper.ReadInConfig()
	if err != nil {
		panic(fmt.Errorf("fatal error while reading config file: %s", err))
	}
}

// createConfiguration uses Viper to initialize a Configuration object.
func createConfiguration() *Configuration {
	configureViper()

	var cookieHashKeyStr = viper.GetString("cookie_hash_key")
	cookieHashKey, err := base64.StdEncoding.DecodeString(cookieHashKeyStr)
	if err != nil {
		panic(fmt.Errorf("Unable to decode cookie hash key \"%s\"",
			cookieHashKeyStr))
	}
	hashLen := len(cookieHashKey)
	if hashLen != 32 && hashLen != 64 {
		panic(fmt.Errorf("Invalid cookie hash key, the length is %d instead of 32/64",
			hashLen))
	}

	var cookieBlockKeyStr = viper.GetString("cookie_block_key")
	cookieBlockKey, err := base64.StdEncoding.DecodeString(cookieHashKeyStr)
	if err != nil {
		panic(fmt.Errorf("Unable to decode cookie block key \"%s\"",
			cookieBlockKeyStr))
	}

	blockLen := len(cookieBlockKey)
	if blockLen != 32 && blockLen != 64 {
		panic(fmt.Errorf("Invalid cookie block key, the length is %d instead of 32/64",
			blockLen))
	}

	return &Configuration{
		ConfigurationFileName: viper.ConfigFileUsed(),
		DatabaseFile:          viper.GetString("database_file"),
		LogFormat:             viper.GetString("log_format"),
		LogLevel:              viper.GetString("log_level"),
		LogOutput:             viper.GetString("log_output"),
		PortNumber:            viper.GetInt("port_number"),
		ReadTimeout:           viper.GetInt64("read_timeout"),
		WriteTimeout:          viper.GetInt64("write_timeout"),
		RepositoryPath:        viper.GetString("repository_path"),
		ServerName:            viper.GetString("server_name"),
		StaticPath:            viper.GetString("static_path"),
		CookieHashKey:         cookieHashKey,
		CookieBlockKey:        cookieBlockKey,
	}
}
